content/blog/2023-12-11-AsyncBril.md

```diff
@@ -0,0 +1,328 @@
+how atomics can be used to manage shared state in a concurrent environment 
+without the need for complex and potentially cumbersome lock-based 
+synchronization mechanisms.
+
+Elaborating more on how our atomics can implement a mutex lock and allow threads
+to concurrently modify an array, here is another useful code snippet: 
+```
+@acquire_lock(lock: atomicint) {
+.loop:
+  res: int = loadatomic lock; 
+  one: int = const 1;
+  eq: bool = eq one res;
+  br eq .loop .acquire_lock;
+.acquire_lock:
+  upd: int = add res one;
+  old: int = cas lock res upd;
+  eq: bool = eq old res;
+  br eq .end .loop;
+  .end:
+  ret;
+}
+
+@release_lock(lock: atomicint) {
+  zero: int = const 0;
+  val: int = swapatomic lock zero;
+}
+```
+
+The `@acquire_lock` function tries to acquire a lock on a shared resource 
+represented by `lock`, an atomic integer. The function first loads the current 
+value of `lock` atomically, and then loops to keep retrying to acquire the lock 
+if it's already being used by another thread. To attempt to acquire the lock, a
+Compare-and-Swap (CAS) operation is used. 
+
+The `@release_lock` function releases the lock. It sets the lock to `0` using 
+an atomic swap operation, indicating that the lock is free. The use of atomic 
+operations ensures that these lock acquire and release operations are 
+thread-safe and prevent race conditions in concurrent environments. 
+
+## Challenges 
+The original implementation of `brilirs` was understandably not designed with 
+multithreading in mind. As a result, we had to change some of the architecture 
+of the data structures and main function headers. To simplify the design, 
+`State` no longer has a parameterized lifetime, and instead uses an atomic 
+reference counter, `Arc<T>`, to hold the program and environment. Further, to 
+implement support for atomic ints, we needed a secondary state variable that
+would all the instantiated atomics. We could not repurpose the heap for this, as 
+`Value` needs to derive the `Copy` trait, but this is not supported for Rust 
+`AtomicI64`. To remedy this, we added an additional data structure, modeled 
+after `Heap`, to store all atomics.  
+
+## Evaluation 
+
+To evaluate our extension, we primarily focused on two key attributes: 
+correctness and runtime. To evaluate the runtime gains from asynchronous 
+programming, consider the following simple benchmark: we use a loop that just 
+adds 1 to a variable. On the first thread it performs 2000000 iterations, on 
+the second thread it performs 1000000, and so on and so forth. One good example
+of when a benchmark like this would be used is in multiplying two large matrices
+together. Below, we show the execution time in miliseconds against the number
+of threads used in the task: 
+
+![Benchmark](eval.jpg)
+
+We observe that from 1 thread to 5 threads, we see substantial performance 
+improvements. This makes sense, as we are taking direct advantage of 
```

Comment:
One thing I don't quite understand about this benchmark: how does it change as you scale up the number of threads? Is the amount of work (total increments) held constant, so when you have N threads you are incrementing the counter, like, 2000000/N times?